### ISSUE TITLE
Syntax highlight breaks down in diffs when ignoring whitespaces

### ISSUE THREAD

**Issue by Sergio0694** [desktop/desktop#14920]
### The problem

When viewing some diffs in a commit that has plenty of whitespace changes (eg. if that commit switched to file-scoped namespaces in C#, which causes each line in the whole file to remove 4 leading whitespaces), the syntax highlight completely breaks down. If you show the whitespaces in the diff, the highlight is correct (though the diff is completely wrong and cluttered), whereas if you hide the whitespaces, the diff is correct but all syntax highlight is completely broken.

### Release version

Version 3.0.3 (x64)

### Operating system

Windows 11 22000.x

### Steps to reproduce the behavior

1. Open a commit with lots of whitespace changes (in C#, for instance)
2. Open a git diff view in the commit history tab
3. In settings, click on "ignore whitespaces"
4. Observe how the syntax highlight breaks down

### Screenshots

A couple of examples:

<img width="560" alt="image" src="https://user-images.githubusercontent.com/10199417/177643998-9c2a31cd-0c58-43ab-9c10-f07063d9a31e.png">

<img width="560" alt="image" src="https://user-images.githubusercontent.com/10199417/177644040-708f9a34-47b9-494b-9e08-ce6ecf97fec5.png">

**Comment by niik** [comment]
Hey @Sergio0694, thanks for reporting this!

Given how the syntax highlighting mechanism in Desktop works (reading the raw contents of the file from the working directory and the raw contents from the Git object database, tokenizing both of them and mashing them together depending on whether the line was added or deleted) I suspect this has always been broken 😞 

I'm sure there's a way to make this works but it will require investigating the white space collapse algorithm in Git a bit more to see if we can reproduce it in the syntax highlighting tokenizer.

cc @say25 was this something you ran into while working on the hide white space feature and do you have any thoughts on approaches to solving this?

cc @sergiou87 @tidy-dev  I'm on the fence about whether to mark this as https://github.com/desktop/desktop/labels/priority-2 but settled on https://github.com/desktop/desktop/labels/priority-3 for now. Thoughts?

**Comment by tsvetilian-ty** [comment]
@niik this issue looks like duplicate of #11681, which is in the backlog 🤔

**Comment by say25** [comment]
I don't think I've run into this exact issue.

I will say, I don't believe Code Mirror (or GitHub.com for that matter) supports some of the new C#9+ features or have different spacing that sometimes is convention. If you are using some of these features in the file it can cause it to choke on more than you would expect.

For example:

```csharp
        public XController
        (
            ILogger<XController> logger,
            IXService xService,
            IYService yService
        )
        {
            _logger = logger.MustNotBeNull();
            _xService = xService.MustNotBeDefault();
            _yService = yService.MustNotBeDefault();
        }
```

vs
 
```csharp
        public XController(ILogger<XController> logger, IXService xService, IYService yService)
        {
            _logger = logger.MustNotBeNull();
            _xService = xService.MustNotBeDefault();
            _yService = yService.MustNotBeDefault();
        }
````

**Comment by Sergio0694** [comment]
I agree that GitHub isn't exactly up to date and I think it's stuck on C# 8 or something (plus it just doesn't work perfectly in some specific cases anyway, such as when using pointers), but the issue here seems to be different to me. As in, even when you're using unsupported syntax for GitHub, some tokens might get the wrong color, but everything is still "consistent". If you collapse whitespaces instead, everything just gets completely mixed up, with individual characters having random colors 🥲

**Comment by niik** [comment]
Yeah, this is definitely due to hiding whitespace. The syntax highlight tokenizer operates on the files with whitespace preserved which is what caused it to get offset like that.

If we knew if a given line had white space removed (and how many characters) we could offset the highlight tokens accordingly. Of the top of my head we could either take the performance penalty of loading the diff both with and without white space or we could attempt to reimplement the white space collapse algorithm in-app. Neither are particularly compelling.

**Comment by niik** [comment]
> @niik this issue looks like duplicate of #11681, which is in the backlog 🤔

Cheers @tsvetilian-ty! Given the ongoing discussion in this I feel we should close that and use this as the canonical issue going forward. I'll take care of it
